Add generalized performance data type in e2e test

Random-Liu opened the pull request:
For kubernetes/contrib/issues/564 and #15554.

This PR added two files in e2e test:
1) `perftype/perftype.go`: This file contains generalized performance data type. The type can be pretty printed in Json format and analyzed by other performance analyzing tools, such as [Perfdash](https://github.com/kubernetes/contrib/tree/master/perfdash).
2) `perf_util.go`: This file contains functions which convert e2e performance test result into new performance data type.

The new performance data type is now used in _Density test, Load test and Kubelet resource tracking_. It's easy to support other e2e performance test by adding new convert function in `perf_util.go`.

@gmarek @yujuhong 
/cc @kubernetes/sig-testing

Random-Liu committed:
Add perfdash support in kubelet_perf and metrics_util

jayunit100 commented on test/e2e/kubelet_perf.go:
```diff
@@ -242,7 +246,10 @@ var _ = Describe("Kubelet [Serial] [Slow]", func() {
 			name := fmt.Sprintf(
-				"for %d pods per node over %v", podsPerNode, monitoringTime)
+				"resource tracking for %d pods per node", podsPerNode)
+			if podsPerNode >= 35 {
```
why 35?

Random-Liu replied:
@jayunit100 There are 3 test cases - "podsPerNode = 0, 35, 100". We only care about the last 2 cases. However, a constant improves the readability. Thanks! :)

gmarek replied:
We run "[Feature:Performance]" in continuous scalability test. I don't think we want to run this in that suite. Maybe "[Feature:NodePerformance]"

Random-Liu replied:
@gmarek ACK. Thanks!

jayunit100 commented:
Cool, I was just thinking about this for the sake of the metrics we create in #22707.

How and where should the Jsons be written during an e2e run?  Is there a standard output_dir for this sort of thing?  Once that is set, we can all use this .

Random-Liu commented:
@jayunit100 Thanks! I'd be very happy if this could help people!

What perf dash does now is quite simple. It periodically gets latest e2e test logs from specified CI jenkins project, parses these logs and generates a multi-line chart of performance metrics vs. build number http://perf-dash.k8s.io/. So you can easily monitor the performance test result and find out in which build a performance regression is induced.

The usage of current version is quite simple ([see here](https://github.com/kubernetes/contrib/pull/565)). You just need to `Logf()` the metrics as `perftype` in your e2e test and configure Perfdash, then it will analyze it automatically.

If this could really help, we could of course do something more to make the PerfDash better. :)

wojtek-t commented on test/e2e/metrics_util.go:
```diff
@@ -284,6 +284,8 @@ func HighLatencyRequests(c *client.Client) (int, error) {
 	}
 
 	Logf("API calls latencies: %s", prettyPrintJSON(metrics))
+	// Log perf data
```
Can you please add a TODO to get rid of the previous line when we migrate to the new form of perf-dash?

Random-Liu replied:
ACK. Thanks, :)

wojtek-t commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+			Data: map[string]float64{
+				"Perc50": float64(apicall.Latency.Perc50) / 1000000, // us -> ms
+				"Perc90": float64(apicall.Latency.Perc90) / 1000000,
+				"Perc99": float64(apicall.Latency.Perc99) / 1000000,
```
I'm not sure we want to have milliseconds. Maybe microseconds?
@gmarek - WDYT?

gmarek replied:
I'd keep milliseconds, especially because we're handling floats here.

wojtek-t commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+		perfData.DataItems = append(perfData.DataItems, item)
+	}
+	return perfData
+}
```
BTW - we should probably change the tests to actually use perftype.PerfData from the begining instead of translating one to the other here.
Can you please add a TODO?

Random-Liu replied:
Sure, I'll add a TODO here. In this PR, I keep the old test result and transform it, so that it could be merged without breaking current perfdash. After the new perfdash is merged, we can start using the new data type from the beginning. :)

gmarek commented on test/e2e/perftype/perftype.go:
```diff
@@ -0,0 +1,65 @@
+const PerfResultTag = "[Result:Performance]"
+
+// String transforms the perfdata into json format with PerfResultTag prefix. If an error occurs
+// , an error message will be returned.
```
move comma to previous line.

Random-Liu replied:
ACK.

gmarek commented on test/e2e/perftype/perftype.go:
```diff
@@ -0,0 +1,65 @@
+
+// String transforms the perfdata into json format with PerfResultTag prefix. If an error occurs
+// , an error message will be returned.
+func (p *PerfData) String() string {
```
Why not simply `prettyPrintJSON`

Random-Liu replied:
@gmarek I just don't wanna to expose the wired `PerfResultTag`. :)

gmarek commented on test/e2e/perftype/perftype.go:
```diff
@@ -0,0 +1,65 @@
+
+// Histogram is a map from bucket to real data point (e.g. "Perc90" -> 23.5). Notice
+// that all data items with the same tag conbination should have the same buckets.
+type Histogram map[string]float64
```
This is not a histogram. Please rename to PerformanceData, PerformanceKeyedData or something along those lines. @davidopp for naming.

jayunit100 replied:
agree, and if it were a true histogram, we maybe could reuse another histogram data type that we already have in the kube deps i.e. prometheus

Random-Liu replied:
ACK. The `Histogram` was there in original Perfdash, I just moved these types here. :)
However, I agree that the name is a little inappropriate. Will rename it. :)

gmarek commented on test/e2e/perftype/perftype.go:
```diff
@@ -0,0 +1,65 @@
+	Data Histogram `json:"data"`
+	// Unit is the data unit. Notice that all data items with the same tag combination
+	// should have the same unit.
+	Unit string `json:"unit"`
```
Maybe embed it into Data structure to make it clear? I.e. make type <NewHistogramName> struct { data map[string]float64, unit string }

gmarek replied:
Oh, I see that DataItem is exactly this struct. Maybe remove this Histogram type and just use the map here?

Random-Liu replied:
ACK. This is also "legacy" of Perfdash. :)

jayunit100 commented:
cc @jeremyeder @timothysc : this is a data type for persisting perf info in e2es , you guys might have feedback.

gmarek commented on test/e2e/perftype/perftype.go:
```diff
@@ -0,0 +1,65 @@
+	// should have the same unit.
+	Unit string `json:"unit"`
+	// Tags is the tags of the data item.
+	Tags map[string]string `json:"tags"`
```
I think that those are essentially Labels. I believe that 'tags' generally mean single strings.

Random-Liu replied:
ACK.

jeremyeder commented:
@jayunit100 if e2e tests output something any tsdb can import, then we're good to go.  json would be perfect.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+	"k8s.io/kubernetes/test/e2e/perftype"
+)
+
+// apiCallToPerfData transforms APIResponsiveness to PerfData. This is used by Density test
```
Remove 'This is used...' sentence.

Random-Liu replied:
ACK.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+	var masterUsage resourceUsagePerContainer
+	var minionsUsage []resourceUsagePerContainer
+	for node, usage := range usagePerNode {
+		if strings.HasSuffix(node, "master") {
```
Use https://github.com/kubernetes/kubernetes/blob/master/pkg/util/system/system_utils.go#L26

Random-Liu replied:
@gmarek Thanks! Reuse is always preferred.

Random-Liu replied:
I found that `IsMasterNode` only receives `api.Node`, which is not applicable here. :)

gmarek replied:
Make it get a Node somehow - we'll probably move from name based identification to label based one in not too distant future.

jayunit100 commented:
I may be crossing wires: but, *how about just using the prometheus data model entirely* ? https://prometheus.io/docs/concepts/data_model/  ?  That way this stuff would work in concert with existing stuff like the PromPush utilities in e2e.

gmarek commented:
@jayunit100 - let's do one thing at a time. If someone will find that JSON is not enough for him/her, we'll think how we can fix it.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+		}
+	}
+	minionAvgUsage := make(resourceUsagePerContainer)
+	for _, minionUsage := range minionsUsage {
```
I'd rather have this perf_util.go hold only simple translation. Please move all computations to kubelet_perf.go

Random-Liu replied:
ACK.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+			if _, found := minionAvgUsage[c]; !found {
+				minionAvgUsage[c] = &containerResourceUsage{Name: usage.Name}
+			}
+			minionAvgUsage[c].CPUUsageInCores += usage.CPUUsageInCores
```
This stands in direct conflict with "this function only cares about memory usage" comment.

Random-Liu replied:
@gmarek Because in kubelet perf, we have more information about cpu usage in `CPUSummry`.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+		for c, usage := range usages {
+			item := perftype.DataItem{
+				Data: map[string]float64{
+					"memory":     float64(usage.MemoryRSSInBytes) / 1000000,
```
s/1000000/1024 * 1024/g

Random-Liu replied:
@gmarek Thanks a lot...This is stupid...Hahaha

jayunit100 commented:
from conversation in slack with @gmarek : 

- we both agree that prometheus already has a data model that solves some of these problems and is battle tested.  
- for now gmarek mentioned there's no need to adopt prometheus' data model,  but we will in the future.

I'd suggest  lets just be conscious of this and maybe even add TODO's in the code that in the long run we want to consolidate on the existing time series data model of prometheus rather than maintain our own.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+		minionAvgUsage[c].MemoryRSSInBytes /= uint64(len(minionsUsage))
+	}
+
+	usageToPerfData := func(usages resourceUsagePerContainer, nodeType string) []perftype.DataItem {
```
Why not top-level function?

Random-Liu replied:
ACK.

gmarek commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,160 @@
+// kubelet_perf now.
+func cpuUsageToPerfData(usagePerNode nodesCPUSummary) *perftype.PerfData {
+	var masterUsage containersCPUSummary
+	var minionsUsage []containersCPUSummary
```
s/minion/node/g

Random-Liu replied:
ACK.

gmarek commented:
I sincerely HATE github review tool. Can we find ANYTHING that lets people reply to other's top level comments? Please? As a response to: https://github.com/kubernetes/kubernetes/pull/22912#issuecomment-196326311
2c: currently we're mainly using two data sources: prometheus for metrics, and cAdvisor for resources. I'd be good to have only one library to handle time series and metrics, but it's an effort that should be started at the bottom of the stack.

jimmidyson commented:
@gmarek I just see these as two different sources of metrics, not in competition with each other but complementing each other. cadvisor exposes Prometheus compatible metrics (using the Prometheus client library). I'm not sure where the conflict is really?

jayunit100 commented:
I think we're all in violent agreement that the Prometheus client is a good baseline for time series and other metric-ish data because of the robust data model.

yujuhong commented:
I pointed @Random-Liu to perf-dash so that we can reuse the existing infrastructure and at least get a clear view to of kubelet resource usage/latency to prevent regression.

I agreed that using Prometheus will be better in the long run (and save us time writing the custom library). In fact, @jimmidyson showed a grafana dashboard in #15554 for this purpose.

jimmidyson commented:
What is it that is considered too complex about the Prometheus data model? Considering everything (most things?) is already instrumented with Prometheus client library it seems to make most sense to just reuse that.

Random-Liu commented:
ack.

gmarek commented:
Nothing in particular - it's just richer and more generic.

yujuhong commented:
@jimmidyson, I don't think there is anything stopping us, other than the amount of work it needs :-)
I'd be happy to see this happen if we can get help or have time to prioritize this.

Random-Liu commented:
Will address the comment soon.

As @yujuhong mentioned above, our original motivation is just to monitor kubelet performance metrics on Perfdash. However, when I saw the code, I found it was impossible to achieve this without refactoring.

@jimmidyson Reusing existing tools like Prometheus is of course better, but it needs a little more effort. :) For now, I think this PR could solve our problem quickly, and I believe it may also help others who need the similar feature recently.

Anyhow, It would be great if we could have a powerful Prometheus dashboard someday!

spiffxp commented:
/cc @kubernetes/sig-scalability

timothysc commented:
> Reusing existing tools like Prometheus is of course better, but it needs a little more effort. :)

We already have hooks into the tests to push via gateway.

yujuhong commented:
> We already have hooks into the tests to push via gateway.

@timothysc, I didn't know that, thanks! Looks like [this file](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/prompush.go) contains the bare-minimum to push the metrics.

@timothysc @jimmidyson @gmarek , should we try consolidating the effort and use Prometheus?

Random-Liu commented:
@gmarek @wojtek-t Addressed the comment, and as @jayunit100 suggested in https://github.com/kubernetes/kubernetes/pull/22912#issuecomment-196326311, added `TODO` to replace the `perftype` with prometheus' data model.

If needed, I'd like to send follow up PRs to switch to Prometheus.

gmarek commented:
@timothysc - I also wasn't aware of this. Let's do this then: #22980

gmarek commented:
The question to @kubernetes/goog-node - do you need this data to be exported and displayed soon, or can this wait for probably ~month?

Random-Liu commented:
@gmarek It is just a nice-to-have feature recently. If we already have a further plan of performance metrics, I think we are fine with waiting. :)

gmarek commented on test/e2e/kubelet_perf.go:
```diff
@@ -242,7 +247,7 @@ var _ = Describe("Kubelet [Serial] [Slow]", func() {
 			podsPerNode := itArg.podsPerNode
 			name := fmt.Sprintf(
-				"for %d pods per node over %v", podsPerNode, monitoringTime)
+				"[Feature:NodePerformance] resource tracking for %d pods per node", podsPerNode)
```
My comment was valid only for ">35" case. For "normal" ones you shouldn't add a Feature tag, as this will disable the test.

Random-Liu replied:
@gmarek Eh, I see. https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e.yaml#L103
I'll remove these "Feature" tags.

Random-Liu replied:
Done. :)

ixdy commented on test/e2e/perf_util.go:
```diff
@@ -0,0 +1,104 @@
+// If an error occurs, nothing will be printed.
+func printPerfData(p *perftype.PerfData) {
+	if str := prettyPrintJSON(p); str != "" {
+		Logf("%s", perftype.PerfResultTag+" "+str)
```
rather than using a tag and grepping the build log, we should be writing to specific performance logs in `testContext.ReportDir` or `testContext.OutputDir`.

Random-Liu replied:
@ixdy, Yeah, it would be better. However, it seems that we have plan to push metrics to Prometheus in the future #22980, so for now I think maybe we could keep the old behavior. :)

gmarek replied:
@Random-Liu +1

gmarek commented on test/e2e/kubelet_stats.go:
```diff
@@ -600,6 +600,40 @@ func (r *resourceMonitor) GetLatest() (resourceUsagePerNode, error) {
+			if _, found := nodeAvgUsage[c]; !found {
+				nodeAvgUsage[c] = &containerResourceUsage{Name: usage.Name}
+			}
+			nodeAvgUsage[c].CPUUsageInCores += usage.CPUUsageInCores
```
It's not a part of this PR, but maybe you or @yujuhong can change the name to 'AverageCPUUsageInCores' or sth - every second week I'm checking if it's cumulative or an average over an interval, and every second week I forget what I learned...

gmarek commented on test/e2e/kubelet_stats.go:
```diff
@@ -600,6 +600,40 @@ func (r *resourceMonitor) GetLatest() (resourceUsagePerNode, error) {
+		}
+	}
+	for c := range nodeAvgUsage {
+		nodeAvgUsage[c].CPUUsageInCores /= float64(len(nodesUsage))
```
This is not true per se, as interval may have different lengths. They shouldn't differ too greatly so it's probably OK to report that. It's good to keep it in mind as soon as we're going to take some actions on such data.

gmarek commented on test/e2e/kubelet_stats.go:
```diff
@@ -661,3 +695,36 @@ func (r *resourceMonitor) GetCPUSummary() nodesCPUSummary {
+	result := make(nodesCPUSummary)
+	var masterSummary containersCPUSummary
+	var nodesSummaries []containersCPUSummary
+	for node, summary := range summaryPerNode {
```
Depending on name will hit us at some point... It's a real pain, as I'm not sure if we can have `map[api.Node]` or do we want to have `map[*api.Node]` in our codebase. In situation like this we'd need some mapping from name to `api.Node` structure which seems like an overkill. @bgrant0607 was opposing depending on names for distinguishing master components from other ones. I don't think there's a big difference between Pods and Nodes in this respect.

Random-Liu replied:
For this problem, in fact there are some other places need to be fixed, just mark here:
```
lantaol@lantaol0:~/workspace/src/k8s.io/kubernetes/test/e2e$ grep -ir "Suffix(node" *
kubelet_stats.go:		if strings.HasSuffix(node, "master") {
kubelet_stats.go:		if strings.HasSuffix(node, "master") {
metrics_grabber_test.go:			if strings.HasSuffix(node.Name, "master") {
metrics_grabber_test.go:			if strings.HasSuffix(node.Name, "master") {
metrics_util.go:		if strings.HasSuffix(node.Name, "master") {
```

gmarek replied:
Note that both `metrics_grabber_test` and `metrics_util` actually have full Node object there, so only kubelet stats needs changes.

gmarek commented:
LGTM, as long as @bgrant0607 won't make move recognizing master Node to use `api.Node` instead of plain node name.

jdef commented:
what about using a Node attribute vs. keying off of the node name?

gmarek commented:
@jdef - I don't understand what you just wrote.

Random-Liu commented:
@gmarek Is there still anything missing in this PR now? :)

gmarek commented:
As long as you're willing to fix the 'suffix' thingy when we move to label-identified HA-master, then no.

Random-Liu commented:
@gmarek Sorry that I was distracted by other things these days. :)
Maybe we could fix the "suffix" related things in another PR? @yujuhong

yujuhong commented:
Currently, there is no information in the node object that we can use to distinguish the master node, except for the name. Could we add a function to encapsulate the `strings.HasSuffix(node.Name, "master")` logic for now, with the "node name" as the input argument? This will make change in the future easier, and we can always modify the test to get the Node object when there is use case for that. @gmarek, are you ok with that?

gmarek commented:
No, but I'm not going to block this PR on this.

Random-Liu commented:
@gmarek Thanks a lot! :)

Random-Liu commented:
@k8s-bot test this issue #21367.

Random-Liu commented:
Self apply LGTM again because I just did a rebase.

gmarek commented:
Yey, we're in! It took some time...
